src/ir.cpp

```diff
@@ -19923,8 +19923,15 @@ static bool ir_analyze_fn_call_generic_arg(IrAnalyze *ira, AstNode *fn_proto_nod
     if (comptime_arg) {
         arg_part_of_generic_id = true;
         arg_val = ir_resolve_const(ira, casted_arg, UndefBad);
-        if (!arg_val)
+        if (!arg_val) {
+            if (casted_arg->value->special == ConstValSpecialRuntime) {
```

Comment:
This is one of the many failure modes of `ir_resolve_const`, the `comptime_arg && !instr_is_comptime(casted_arg)` check makes more sense before trying to resolve the const value.

The PR needs a test case in `compile-errors`.

Reply:
Like this?

```zig
if (comptime_arg) {
        if (!instr_is_comptime(casted_arg)) {
            ir_add_error(ira, &casted_arg->base,
                buf_sprintf("runtime value cannot be passed to comptime arg"));
            return false;
        }
        arg_part_of_generic_id = true;
        arg_val = ir_resolve_const(ira, casted_arg, UndefBad);
        if (!arg_val)
            return false;
    } else {
// ...
```

Reply:
yep, you can pull it out of the `if (comptime_arg)` (and add `comptime_arg && ...` to the check) if you prefer less indentation.
Functionality-wise it should be equivalent to your original check.

Reply:
Thanks, updated now. 

I just changed the error message for the `generic function instance with non-constant expression` test since it does the same check. Seems like the CI is failing due to llvm 11.

Reply:
The else in `else if (comptime_arg)` is wrong, you want to always execute `ir_resolve_const` if `comptime_arg` is true. When `instr_is_comptime` is false you quit the function and raise an error.

> Seems like the CI is failing due to llvm 11.

Weird, are you sure the master you branched off is recent enough?